torch/_functorch/eager_transforms.py

```diff
@@ -1279,112 +1283,14 @@ def wrapper(*args, **kwargs):
-        >>> assert torch.allclose(neg_sin_x, -x.sin())
-
-    When composed with ``vmap``, ``grad`` can be used to compute per-sample-gradients:
-
-        >>> # xdoctest: +SKIP
-        >>> from torch.func import grad, vmap
-        >>> batch_size, feature_size = 3, 5
-        >>>
-        >>> def model(weights, feature_vec):
-        >>>     # Very simple linear model with activation
-        >>>     assert feature_vec.dim() == 1
-        >>>     return feature_vec.dot(weights).relu()
-        >>>
-        >>> def compute_loss(weights, example, target):
-        >>>     y = model(weights, example)
-        >>>     return ((y - target) ** 2).mean()  # MSELoss
-        >>>
-        >>> weights = torch.randn(feature_size, requires_grad=True)
-        >>> examples = torch.randn(batch_size, feature_size)
-        >>> targets = torch.randn(batch_size)
-        >>> inputs = (weights, examples, targets)
-        >>> grad_weight_per_example = vmap(grad(compute_loss), in_dims=(None, 0, 0))(*inputs)
-
-    Example of using ``grad`` with ``has_aux`` and ``argnums``:
-
-        >>> # xdoctest: +SKIP
-        >>> from torch.func import grad
-        >>> def my_loss_func(y, y_pred):
-        >>>    loss_per_sample = (0.5 * y_pred - y) ** 2
-        >>>    loss = loss_per_sample.mean()
-        >>>    return loss, (y_pred, loss_per_sample)
-        >>>
-        >>> fn = grad(my_loss_func, argnums=(0, 1), has_aux=True)
-        >>> y_true = torch.rand(4)
-        >>> y_preds = torch.rand(4, requires_grad=True)
-        >>> out = fn(y_true, y_preds)
-        >>> # > output is ((grads w.r.t y_true, grads w.r.t y_preds), (y_pred, loss_per_sample))
-
-    .. note::
-        Using PyTorch ``torch.no_grad`` together with ``grad``.
-
-        Case 1: Using ``torch.no_grad`` inside a function:
-
-            >>> # xdoctest: +SKIP
-            >>> def f(x):
-            >>>     with torch.no_grad():
-            >>>         c = x ** 2
-            >>>     return x - c
-
-        In this case, ``grad(f)(x)`` will respect the inner ``torch.no_grad``.
-
-        Case 2: Using ``grad`` inside ``torch.no_grad`` context manager:
-
-            >>> # xdoctest: +SKIP
-            >>> with torch.no_grad():
-            >>>     grad(f)(x)
-
-        In this case, ``grad`` will respect the inner ``torch.no_grad``, but not the
-        outer one. This is because ``grad`` is a "function transform": its result
-        should not depend on the result of a context manager outside of ``f``.
-
-    """
-    @wraps(func)
-    def wrapper(*args, **kwargs):
-        results = grad_and_value(func, argnums, has_aux=has_aux)(*args, **kwargs)
-        if has_aux:
-            grad, (_, aux) = results
-            return grad, aux
-        grad, _ = results
-        return grad
-    return wrapper
-
+def grad_impl(func: Callable, argnums: argnums_t, has_aux: bool, args, kwargs):
+    func = lazy_dynamo_disable(func)
+    results = grad_and_value(func, argnums, has_aux=has_aux)(*args, **kwargs)
```

Comment:
We should really make grad_and_value_impl the higher-order operator, so that we get support for both grad and grad_and_value. Doesn't need to happen in this PR.